packages/ai/core/tool/mcp.ts

```diff
@@ -0,0 +1,227 @@
+import { StdioClientTransport } from '@modelcontextprotocol/sdk/client/stdio.js';
+import { SSEClientTransport } from '@modelcontextprotocol/sdk/client/sse.js';
+
+import { jsonSchema, Schema } from '@ai-sdk/ui-utils';
+import { AISDKError, JSONSchema7 } from '@ai-sdk/provider';
+
+import { ToolSet } from './../generate-text/tool-set';
+import { inferParameters, Tool, tool } from './tool';
+import { z } from 'zod';
+import { IOType } from 'node:child_process';
+import { Stream } from 'node:stream';
+import { SimpleMcpClient } from './mcp/client';
+import {
+  CallToolRequest,
+  CallToolResult,
+  CallToolResultSchema,
+  CompatibilityCallToolResultSchema,
+  Implementation,
+  ListToolsRequest,
+  ListToolsResult,
+  RequestOptions,
+} from './mcp/types';
+
+interface McpStdioServerConfig {
+  command: string;
+  args?: string[];
+  env?: Record<string, string>;
+  stderr?: IOType | Stream | number;
+  cwd?: string;
+  type: 'stdio';
+}
+interface McpSSEServerConfig {
+  type: 'sse';
+  url: string;
+  options?: {
+    // authProvider?: OAuthClientProvider;
+    eventSourceInit?: EventSourceInit;
+    requestInit?: RequestInit;
+  };
+}
+
+const DEFAULT_CLIENT_CONFIG = {
+  name: 'ai-sdk-mcp-client',
+  version: '1.0.0',
+  connectionTimeoutMs: 6000,
+  requestTimeoutMs: 3000,
+};
+
+/**
+ * An interface that defines the minimum required MCP client properties.
+ * Used when endusers provide their own MCP Client for more granular control.
+ */
+interface McpClientInterface {
+  listTools(
+    params?: ListToolsRequest['params'],
+    options?: RequestOptions,
+  ): Promise<ListToolsResult>;
+  callTool(
+    params: CallToolRequest['params'],
+    resultSchema?:
+      | typeof CallToolResultSchema
+      | typeof CompatibilityCallToolResultSchema,
```

Comment:
why 2?

Reply:
copied over directly from `mpc` ts sdk! seems like this was due to supporting backwards compatibility